fluentinterface/src/main/java/com/iluwatar/fluentinterface/fluentiterable/lazy/DecoratingIterator.java

```diff
@@ -0,0 +1,53 @@
+package com.iluwatar.fluentinterface.fluentiterable.lazy;
+
+import java.util.Iterator;
+
+/**
+ * This class is used to realize LazyFluentIterables. It decorates
+ * a given iterator.
+ * @param <TYPE>
+ */
+public abstract class DecoratingIterator<TYPE> implements Iterator<TYPE> {
+
+    protected final Iterator<TYPE> fromIterator;
+
+    private TYPE next = null;
+
+    /**
+     * Creates an iterator that decorates the given iterator.
+     * @param fromIterator
+     */
+    public DecoratingIterator(Iterator<TYPE> fromIterator) {
+        this.fromIterator = fromIterator;
+    }
+
+    /**
+     * Precomputes and caches the next element of the iteration.
+     * @return true if a next element is available
+     */
+    @Override
+    public final boolean hasNext() {
+        next = computeNext();
```

Comment:
I can't see a way how to not request the next element, in order to figure out if there's a next one or not with stacked iterators. A fluent iterator with an additional action has to apply this action onto the next element, so it has to be computed. I think that's also the way guava works internally, or did I miss the point?